File: src/console/command/parameter/def/MultipleTypesParameter.php
<?php

namespace pocketcloud\cloud\console\command\parameter\def;

use pocketcloud\cloud\console\command\parameter\CommandParameter;
use pocketcloud\cloud\console\command\parameter\exception\ArgumentParseException;

readonly class MultipleTypesParameter extends CommandParameter {

    /**
     * @param string $name
     * @param array<CommandParameter> $allowedTypes
     * @param bool $optional
     * @param string|null $customErrorMessage
     */
    public function __construct(
        string $name,
        protected array $allowedTypes,
        bool $optional,
        ?string $customErrorMessage = null
    ) {
        parent::__construct($name, $optional, $customErrorMessage);
    }

    public function parseValue(string $input): mixed {
        $result = null;
        foreach ($this->allowedTypes as $type) {
            try {
                $result = $type->parseValue($input);
                break;
            } catch (ArgumentParseException) {
                continue;
            }
        }

        return $result ?? throw new ArgumentParseException();
    }

    public function onTabCompleteMatch(string $currentArg): array {
        $matches = [];
        foreach ($this->allowedTypes as $type) {
            $matches = array_merge($type->onTabCompleteMatch($currentArg), $matches);
        }

        return array_unique(array_values($matches));
    }

    public function getType(): string {
        return implode("|", array_map(fn(CommandParameter $argument) => $argument->getName(), $this->allowedTypes));
    }

    public function getAllowedTypes(): array {
        return $this->allowedTypes;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/console/command/parameter/def/MultipleTypesParameter.java
package de.pocketcloud.cloud.console.command.parameter.def;

import de.pocketcloud.cloud.console.command.parameter.BaseCommandParameter;
import de.pocketcloud.cloud.console.command.exception.ArgumentParseException;
import lombok.Getter;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Getter
public class MultipleTypesParameter extends BaseCommandParameter {

    private final List<BaseCommandParameter> allowedTypes;

    public MultipleTypesParameter(String name, List<BaseCommandParameter> allowedTypes, boolean optional, String customErrorMessage) {
        super(name, optional, customErrorMessage);
        this.allowedTypes = List.copyOf(allowedTypes);
    }

    public MultipleTypesParameter(String name, boolean optional, BaseCommandParameter... allowedTypes) {
        this(name, Arrays.asList(allowedTypes), optional, null);
    }

    @Override
    public Object parseValue(String input) throws ArgumentParseException {
        for (BaseCommandParameter type : allowedTypes) {
            try {
                return type.parseValue(input);
            } catch (ArgumentParseException ignored) {}
        }
        throw new ArgumentParseException();
    }

    @Override
    public List<String> onTabCompleteMatch(String currentArg) {
        List<String> combined = new ArrayList<>();
        for (BaseCommandParameter type : allowedTypes) combined.addAll(type.onTabCompleteMatch(currentArg));
        return combined.stream().distinct().collect(Collectors.toList());
    }

    @Override
    public String getType() {
        return allowedTypes.stream().map(BaseCommandParameter::getName).collect(Collectors.joining("|"));
    }
}